docs/libvirt-pod-networking.md

```diff
@@ -0,0 +1,82 @@
+# Using CNI Interfaces and libvirt for networking
+
+Author: Fabian Deutsch \<fabiand@redhat.com\>
+
+## Introduction
+
+Today networking in KubeVirt is not integrated with Kubernetes, a user can
+rather choose to create a specific VM Spec definition to create a particular
+libvirt domxml, to reuse an existing physical device.
+But this is obviously suboptimal
+
+- It is not integrated with Kubernetes
+- It does not provide connectivity with pods
+- It requires to specify host specific bits (the network device to use)
+
+The purpose of this document is to describe one approach of how we can provide
+a minimal integrated solution, with few workarounds.
+
+
+## Use-case
+
+The primary use-case is to allow VM to get connected to the pod network.
+
+
+## API
+
+The API is pretty clear: As usual an interface will be defined.
+To indicate that Pod interfaces shall be used for networking, a new type
+`pod-network` is introduced.
+All other fields, concerning the properties of the virtual NIC, can be used
+as usual.
+
+```yaml
+kind: VM
+spec:
+  domain:
+    devices:
+      interfaces:
+      - type: pod-network
+        mac:
+          address: 00:11:22:33:44:55
+        model:
+          type: virtio
+        source:
+          network: default
+```
+
+The `source` field can be omitted initially, as Kubernetes only provides a
+single network for Pods today.
+In future we might want to allow it to select the Pod network to connect to.
+
+
+## Implementation
+
+The broader steps to enable networking in the described way:
+
+1. Request/Create a new Pod interface on the libvirt pod
+2. Remember the IP of the Pod interface, then remove it from the Pod interface
+3. Create a libvirt network, backed by a bridge, enable DHCP,
+   and add a single host entry, with the IP address of the Pod, and the MAC
+   address of the corresponding virtual NIC of the VM
```

Comment:
Hm, @fabiand, I think you leave the Pod IP of the libvirt pod alone. It can be used by libvirt/qemu for spice, ...

To get new IPs for VMs, virt-handler or something on that layer would call CNI directly, to add a new device the the libvirt Pod namespace. Then you use this IP and send the traffic through the device, created by CNI for this IP. That should pretty much overcome the problems mentioned .

virt-handler can also shutdown the devices via CNI, after the VM stopped, so that the IP is returned to the provider pool. The question is, how we can weave that properly into the Pod shutdown process.